x/core/keeper/msg_server.go

```diff
@@ -52,6 +55,100 @@ func (m msgServer) AddToAllowlist(goCtx context.Context, msg *types.MsgAddToAllo
 	return &types.MsgAddToAllowlistResponse{}, nil
 }
 
+func (m msgServer) Stake(goCtx context.Context, msg *types.MsgStake) (*types.MsgStakeResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Verify stake proof.
+	var sequenceNum uint64
+	var isExistingStaker bool
+	staker, err := m.Stakers.Get(ctx, msg.PublicKey)
+	if err != nil {
+		if !errors.Is(err, collections.ErrNotFound) {
+			return nil, err
+		}
+	} else {
+		sequenceNum = staker.SequenceNum
+		isExistingStaker = true
+	}
+
+	hash, err := msg.ComputeStakeHash("", ctx.ChainID(), sequenceNum)
+	if err != nil {
+		return nil, err
+	}
+	publicKey, err := hex.DecodeString(msg.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+	proof, err := hex.DecodeString(msg.Proof)
+	if err != nil {
+		return nil, err
+	}
+	_, err = vrf.NewK256VRF().Verify(publicKey, proof, hash)
```

Comment:
So in the contract/rust-common we have one statically created instance of the VRF struct and then use it when necessary instead of creating a new one every time.

Not for this PR, and that might have mattered more for WASM world than not, but might be worth considering at some point since that type of math can be expensive to re-instantiate every time.